discovery/consul/consul.go

```diff
@@ -226,92 +286,138 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 		default:
 		}
 
+		// Get the local datacenter first, if necessary.
+		err := d.getDatacenter()
 		if err != nil {
-			level.Error(d.logger).Log("msg", "Error refreshing service list", "err", err)
-			rpcFailuresCount.Inc()
 			time.Sleep(retryInterval)
 			continue
 		}
-		// If the index equals the previous one, the watch timed out with no update.
-		if meta.LastIndex == lastIndex {
-			continue
-		}
-		lastIndex = meta.LastIndex
 
-		// If the datacenter was not set from clientConf, let's get it from the local Consul agent
-		// (Consul default is to use local node's datacenter if one isn't given for a query).
-		if d.clientDatacenter == "" {
-			info, err := d.client.Agent().Self()
-			if err != nil {
-				level.Error(d.logger).Log("msg", "Error retrieving datacenter name", "err", err)
-				time.Sleep(retryInterval)
-				continue
+		if len(d.watchedServices) == 0 || d.watchedTag != "" {
+			// We need to watch the catalog.
+			d.watchServices(ctx, ch, &lastIndex, services)
+		} else {
+			// We only have fully defined services.
+			for _, name := range d.watchedServices {
+				ctx, _ := context.WithCancel(ctx)
+				d.watchService(name, ctx, ch)
 			}
-			d.clientDatacenter = info["Config"]["Datacenter"].(string)
+			// Wait for cancellation.
+			<-ctx.Done()
+			break
 		}
+	}
+}
 
-		// Check for new services.
-		for name := range srvs {
-			if !d.shouldWatch(name) {
-				continue
-			}
-			if _, ok := services[name]; ok {
-				continue // We are already watching the service.
-			}
-
-			srv := &consulService{
-				client: d.client,
-				name:   name,
-				labels: model.LabelSet{
-					serviceLabel:    model.LabelValue(name),
-					datacenterLabel: model.LabelValue(d.clientDatacenter),
-				},
-				tagSeparator: d.tagSeparator,
-				logger:       d.logger,
-			}
-
-			wctx, cancel := context.WithCancel(ctx)
-			go srv.watch(wctx, ch)
+// Watch the catalog for services we would like to watch. This is only necessary
+// for services filtered by tag. If we already have the name it is more efficient to
+// watch nodes the service directly.
```

Comment:
this comment is a bit unclear imo (unless I'm misunderstanding what this function is doing), can we change it to something like `Watch the catalog for changes in instances of the services we would like to watch. Adds/removes instances of the service found in the catalog to Discovery's targets.`?

Reply:
`service instance` (node?) is unclear too. I changed it to:

```
// Watch the catalog for new services we would like to watch. This is called only                                                                                                                                 
// when we don't know yet the names of the services and need to ask Consul the                                                                                                                                    
// entire list of services.   
```

Reply:
Hmmm. Okay, now I think I just misunderstood what's supposed to happen here. This catches the case where we want to watch a service that consul doesn't yet know about, which it then finds later. 👍